### ISSUE TITLE
[FEAT]:  Flowise APi

### ISSUE THREAD

**Issue by SerenityNrrd** [Mintplex-Labs/anything-llm#1785]
### What would you like to see?

Do you have a Template to create a Custom Endpoint for your AnythingLLM ? I want to create a way to use Flowise API https://docs.flowiseai.com/using-flowise/api - and have it work with your AnyhtingLLM framework. ?

**Comment by SerenityNrrd** [comment]
I wrote this code. import os
import time
import logging
import json
from flask import Flask, request, jsonify, Response, stream_with_context
import requests
from typing import Dict, Any

# Set up logging
logging.basicConfig(level=logging.DEBUG, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

app = Flask(__name__)

# Configuration
FLOWISE_API_URL = os.getenv('FLOWISE_API_URL', 'http://localhost:3333/api/v1/prediction')
FLOWISE_API_KEY = os.getenv('FLOWISE_API_KEY', 'secret-key')
CHATFLOW_ID = os.getenv('CHATFLOW_ID', 'b6f12e2a-49e3-4930-a19a-500b1179ddd9')

# Log environment variables to verify they are set correctly
logger.debug(f"FLOWISE_API_URL: {FLOWISE_API_URL}")
logger.debug(f"FLOWISE_API_KEY: {FLOWISE_API_KEY}")
logger.debug(f"CHATFLOW_ID: {CHATFLOW_ID}")

# Model information
GETUPD8_MODEL = {
    "id": "Getupd8",
    "object": "model",
    "created": int(time.time()),
    "owned_by": "getupd8-owner"
}

SECOND_MODEL = {
    "id": "SecondModel",
    "object": "model",
    "created": int(time.time()),
    "owned_by": "secondmodel-owner"
}

def convert_to_flowise_format(data: Dict[str, Any]) -> Dict[str, Any]:
    """Convert OpenAI format to Flowise format for both chat and completion."""
    logger.debug(f"Input data: {data}")
    
    if 'messages' in data:
        # Chat format
        question = next((msg['content'] for msg in reversed(data['messages']) if msg['role'] == 'user'), '')
        history = [
            {'role': 'apiMessage' if msg['role'] == 'assistant' else 'userMessage', 'content': msg['content']}
            for msg in data['messages'][:-1]  # Exclude the last user message as it's the current question
        ]
    else:
        # Completion format
        question = data.get('prompt', '')
        history = []

    flowise_data = {
        'question': question,
        'history': history
    }
    logger.debug(f"Converted Flowise format: {flowise_data}")
    return flowise_data

def convert_to_openai_format(flowise_response: Dict[str, Any], is_chat: bool) -> Dict[str, Any]:
    """Convert Flowise response to OpenAI format for both chat and completion."""
    logger.debug(f"Flowise response: {flowise_response}")
    
    # Extract the assistant's response from the Flowise response
    assistant_message = flowise_response.get('text', '')
    
    if is_chat:
        openai_format = {
            'id': 'chatcmpl-' + os.urandom(4).hex(),
            'object': 'chat.completion',
            'created': int(time.time()),
            'model': 'Getupd8',
            'choices': [{
                'index': 0,
                'message': {
                    'role': 'assistant',
                    'content': assistant_message
                },
                'finish_reason': 'stop'
            }],
            'usage': {
                'prompt_tokens': -1,
                'completion_tokens': -1,
                'total_tokens': -1
            }
        }
    else:
        openai_format = {
            'id': 'cmpl-' + os.urandom(4).hex(),
            'object': 'text_completion',
            'created': int(time.time()),
            'model': 'Getupd8',
            'choices': [{
                'text': assistant_message,
                'index': 0,
                'logprobs': None,
                'finish_reason': 'stop'
            }],
            'usage': {
                'prompt_tokens': -1,
                'completion_tokens': -1,
                'total_tokens': -1
            }
        }
    
    logger.debug(f"Converted OpenAI format: {openai_format}")
    return openai_format

def process_request(data: Dict[str, Any], is_chat: bool):
    logger.info(f"Processing {'chat' if is_chat else 'completion'} request")
    
    model_id = data.get('model', 'Getupd8')
    
    if model_id == SECOND_MODEL['id']:
        # Return a static response for the second model
        static_response = {
            'id': 'chatcmpl-' + os.urandom(4).hex(),
            'object': 'chat.completion',
            'created': int(time.time()),
            'model': SECOND_MODEL['id'],
            'choices': [{
                'index': 0,
                'message': {
                    'role': 'assistant',
                    'content': 'This is a static response from the second model.'
                },
                'finish_reason': 'stop'
            }],
            'usage': {
                'prompt_tokens': -1,
                'completion_tokens': -1,
                'total_tokens': -1
            }
        }
        return jsonify(static_response)
    
    flowise_data = convert_to_flowise_format(data)
    headers = {
        'Content-Type': 'application/json',
        'Authorization': f'Bearer {FLOWISE_API_KEY}'
    }

    stream = data.get('stream', False)
    logger.debug(f"Stream mode: {stream}")

    try:
        if stream:
            def generate():
                with requests.post(f"{FLOWISE_API_URL}/{CHATFLOW_ID}", json=flowise_data, headers=headers, stream=True) as response:
                    logger.debug(f"Streaming response from Flowise API")
                    assistant_message = ""
                    for line in response.iter_lines():
                        if line:
                            decoded_line = line.decode('utf-8')
                            logger.debug(f"Received line: {decoded_line}")
                            flowise_response = json.loads(decoded_line)
                            assistant_message += flowise_response.get('text', '') + "\n"
                            openai_response = convert_to_openai_format(flowise_response, is_chat)
                            yield f"data: {json.dumps(openai_response)}\n\n"
                    final_response = {
                        'choices': [{
                            'index': 0,
                            'message': {
                                'role': 'assistant',
                                'content': assistant_message.strip()
                            },
                            'finish_reason': 'stop'
                        }],
                        'usage': {
                            'prompt_tokens': -1,
                            'completion_tokens': -1,
                            'total_tokens': -1
                        }
                    }
                    yield f"data: {json.dumps(final_response)}\n\n"
                    yield "data: [DONE]\n\n"
            return Response(stream_with_context(generate()), content_type='text/event-stream')
        else:
            logger.debug(f"Sending request to Flowise API: {flowise_data}")
            response = requests.post(f"{FLOWISE_API_URL}/{CHATFLOW_ID}", json=flowise_data, headers=headers)
            response.raise_for_status()  # Raise an exception for HTTP errors
            logger.debug(f"Received response from Flowise API: {response.text}")
            flowise_response = response.json()
            openai_response = convert_to_openai_format(flowise_response, is_chat)
            logger.debug(f"Converted OpenAI response: {openai_response}")
            return jsonify(openai_response)
    except requests.exceptions.RequestException as e:
        logger.error(f"Request to Flowise API failed: {e}")
        return jsonify({"error": "Request to Flowise API failed"}), 500

@app.route('/v1/chat/completions', methods=['POST'])
def chat_completions():
    logger.info("Received request to /v1/chat/completions")
    return process_request(request.json, is_chat=True)

@app.route('/v1/completions', methods=['POST'])
def completions():
    logger.info("Received request to /v1/completions")
    return process_request(request.json, is_chat=False)

@app.route('/v1/models', methods=['GET'])
def list_models():
    logger.info("Received request to list models")
    return jsonify({
        "object": "list",
        "data": [GETUPD8_MODEL, SECOND_MODEL],
    })

@app.route('/v1/models/<model_id>', methods=['GET'])
def retrieve_model(model_id):
    logger.info(f"Received request to retrieve model: {model_id}")
    if model_id == GETUPD8_MODEL['id']:
        return jsonify(GETUPD8_MODEL)
    elif model_id == SECOND_MODEL['id']:
        return jsonify(SECOND_MODEL)
    else:
        return jsonify({"error": "Model not found"}), 404

if __name__ == '__main__':
    logger.info("Starting Flask application")
    app.run(debug=True, port=3334)
--- trying to make it work but when I run it... i get the respose back from Flowise 2024-06-28 15:31:28,096 - DEBUG - Flowise response: {'text': 'Hello! How can I assist you today?', 'question': 'Hello', 'chatId': 'be9c0230-5b8b-452c-b562-34657f75aed8', 'chatMessageId': 'b200060f-2b1c-4773-a08c-fef921ba5d1a', 'sessionId': 'be9c0230-5b8b-452c-b562-34657f75aed8'}
2024-06-28 15:31:28,097 - DEBUG - Converted OpenAI format: {'id': 'chatcmpl-59dde35d', 'object': 'chat.completion', 'created': 1719603088, 'model': 'Getupd8', 'choices': [{'index': 0, 'message': {'role': 'assistant', 'content': 'Hello! How can I assist you today?'}, 'finish_reason': 'stop'}], 'usage': {'prompt_tokens': -1, 'completion_tokens': -1, 'total_tokens': -1}}
2024-06-28 15:31:28,097 - INFO - 127.0.0.1 - - [28/Jun/2024 15:31:28] "POST /v1/chat/completions HTTP/1.1" 200 - --- but nothing shows up in the Chat 

![image](https://github.com/Mintplex-Labs/anything-llm/assets/6841787/d4443fba-61d0-410c-a1cb-012faea73e34)

**Comment by SerenityNrrd** [comment]
but when it runs nothing  ends up in the chat window
![image](https://github.com/Mintplex-Labs/anything-llm/assets/6841787/82ae9c5a-8fae-4181-a586-134cd5787e93)
```2024-06-28 15:31:28,096 - DEBUG - Flowise response: {'text': 'Hello! How can I assist you today?', 'question': 'Hello', 'chatId': 'be9c0230-5b8b-452c-b562-34657f75aed8', 'chatMessageId': 'b200060f-2b1c-4773-a08c-fef921ba5d1a', 'sessionId': 'be9c0230-5b8b-452c-b562-34657f75aed8'}
2024-06-28 15:31:28,097 - DEBUG - Converted OpenAI format: {'id': 'chatcmpl-59dde35d', 'object': 'chat.completion', 'created': 1719603088, 'model': 'Getupd8', 'choices': [{'index': 0, 'message': {'role': 'assistant', 'content': 'Hello! How can I assist you today?'}, 'finish_reason': 'stop'}], 'usage': {'prompt_tokens': -1, 'completion_tokens': -1, 'total_tokens': -1}}
2024-06-28 15:31:28,097 - INFO - 127.0.0.1 - - [28/Jun/2024 15:31:28] "POST /v1/chat/completions HTTP/1.1" 200 -

**Comment by timothycarambat** [comment]
Are you using streaming? I presume you are using the Generic OpenAI connector for chatting?

**Comment by SerenityNrrd** [comment]
yes, I am trying to use the generic openAI connector for now but i would love to make a custom connector so we dont have to do this weird convertion.

**Comment by timothycarambat** [comment]
But you end goal is basically just to connect to some flowise process endpoint to process a chat, correct?

**Comment by SerenityNrrd** [comment]
Yes, Specially I would like to use Anything as the front end for Flowise and then use a few different flows as Models.

**Comment by SerenityNrrd** [comment]
Basicially I can get the response out.. but no matter what I try i can't get AnythingLLM to show the response from Flowise.

**Comment by SerenityNrrd** [comment]
> Are you using streaming? I presume you are using the Generic OpenAI connector for chatting?

Thanks to this Question i was able to Figure out my issue and got it working.

**Comment by timothycarambat** [comment]
I think issue for an integration with flowise still makes sense though

**Comment by SerenityNrrd** [comment]
> I think issue for an integration with flowise still makes sense though

I completely agree I searched for hours there is NO frontend GUI offering it yet. (hence we trying to figure it out myself). Flowise allows anyone to make Complete Flows that would take hours to try to implement on a coding end. My use case is to be able to assign a few different Flows as "Models" so that Clients can Choose what Function or Information path they want to follow. for instance . Choose Fee Calcuation and it will Do any Fee Calcuations and focus on that or if they want to Ask questions about a Certain Chapter.. it would specialize in that chapter. and since Flowise now offers Agents and Flows like that the possiblity is pretty endless what you could build with it.. and yet it lacks a good frontend with user control and Various other things.

**Comment by nickp27** [comment]
> > Are you using streaming? I presume you are using the Generic OpenAI connector for chatting?
> 
> Thanks to this Question i was able to Figure out my issue and got it working.

What was the solution? Do you mind sharing? There doesn't seem to be an obvious guide for converting an Flowise endpoint to an openai compatible endpoint.

**Comment by SerenityNrrd** [comment]
> > > Are you using streaming? I presume you are using the Generic OpenAI connector for chatting?
> > 
> > 
> > Thanks to this Question i was able to Figure out my issue and got it working.
> 
> What was the solution? Do you mind sharing? There doesn't seem to be an obvious guide for converting an Flowise endpoint to an openai compatible endpoint.

I shared the python script that i wrote in the discord. https://discord.com/channels/1114740394715004990/1248104856732827658 but mainly the solution was to create a middlewear server that just takes the OPENAI compatible input run the Flowise api and convert it back to openai. but it seems to be working on the lowest level but hopefully eventually we will get a full use solution in the near future.

**Comment by timothycarambat** [comment]
If you have a flowise execution endpoint you wish to leverage you can now set that as a single API block in a new Agent flow and then have it available as a tool - depending on what your flowise tool looks like of course.

**Comment by zfusx** [comment]
we wrote a serve to bride anythingllm and flowise ai , 
"A middleware server that creates an OpenAI-compatible API layer for Flowise, allowing you to use Flowise chatflows with any application or library designed to work with OpenAI's chat completions API."
anyone would struggle with link anythingllm to flowise ai ,can have a try:
https://github.com/zfusx/toble-bridge
